### PR TITLE
PR: run-time check: verify if wrong conda/python usage during windows build

### PR DESCRIPTION
**Pull request by PeterDSteinberg**
This checks to see if perhaps the chalmers / anaonda-build configuration on windows will result in conda trying to modify the running conda/python executable which isn't allowed on windows.

### CODE REVIEW

File: binstar_build_client/worker/worker.py [hunk 1]
```diff
@@ -390,3 +390,22 @@ def job_context(self, journal, job_data):
...
+    @classmethod
+    def check_conflicting_procs(cls):
+        if os.name != 'nt':
+            return
+        procs_on_wrong_python = []
+        executable_dir = get_conda_root_prefix()
+        for proc in psutil.process_iter():
+            try:
+                cmd = proc.cmdline()
+            except psutil.AccessDenied:
```
Comment (srossross) [inline binstar_build_client/worker/worker.py]: why does this happen?

File: binstar_build_client/worker/worker.py [hunk 2]
```diff
@@ -390,3 +390,22 @@ def job_context(self, journal, job_data):
...
 
+    @classmethod
+    def check_conflicting_procs(cls):
+        if os.name != 'nt':
+            return
+        procs_on_wrong_python = []
+        executable_dir = get_conda_root_prefix()
+        for proc in psutil.process_iter():
+            try:
+                cmd = proc.cmdline()
```
Comment (srossross) [inline binstar_build_client/worker/worker.py]: I think you want `proc.exe()`

File: binstar_build_client/worker/worker.py [hunk 3]
```diff
@@ -390,3 +390,22 @@ def job_context(self, journal, job_data):
             duration = time.time() - start_time
             log.info('Build Duration {0} seconds'.format(duration))
 
+    @classmethod
+    def check_conflicting_procs(cls):
```
Comment (srossross) [inline binstar_build_client/worker/worker.py]: can you move this function to a new file?

**Comment by srossross** [comment]
@PeterDSteinberg can you post your results here? You posted them in the private flowdoc chat.

**Comment by PeterDSteinberg** [comment]
Note that the test on the second commit to this PR failed because it said `anaconda worker` was running on the same Python as the builds.  That is an interesting failure, confirming what the test was supposed to catch.  Oddly, other [windows builds recenlty](https://anaconda.org/binstar/binstar-build/builds/419) have been okay, despite apparently running `anaconda worker run` and builds in the same Python.

**Comment by srossross** [comment]
> Oddly, other windows builds recently have been okay, despite apparently running anaconda worker run and builds in the same Python

@PeterDSteinberg why is this odd? The error only occurs when the build script executes `conda create` and tries to update Python. The unit-tests tests do not run this command.

**Comment by PeterDSteinberg** [comment]
@srossross It looks like [this successful build](https://anaconda.org/binstar/binstar-build/builds/419/4#L52) used `conda create` on windows in the last few days.

**Comment by srossross** [comment]
@PeterDSteinberg Can you please clarify. You are confusing me because I am not sure what is a `test` failure vs what is failing because of fundamental flaws in the build worker.

**Comment by srossross** [comment]
For example this is a test failure. Is this what you were referring to above? If so. I would expect this to fail, because that is what you wrote the code to do:

```
======================================================================
ERROR: test_build_timeout (binstar_build_client.worker.tests.test_jobs.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\binstar\builds\binstar\binstar-build\env\lib\site-packages\mock\mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "C:\Users\binstar\builds\binstar\binstar-build\source\binstar_build_client\worker\tests\test_jobs.py", line 238, in test_build_timeout
    failed, status = worker.build(job_data)
  File "C:\Users\binstar\builds\binstar\binstar-build\source\binstar_build_client\worker\worker.py", line 234, in build
    validate_procs()
  File "C:\Users\binstar\builds\binstar\binstar-build\source\binstar_build_client\worker\utils\validate_procs.py", line 34, in validate_procs
    "Python prefix: {}".format(" ".join(procs_on_wrong_python)))

BinstarError: There were processes running on the incorrect Python prefix: Pid 1408 is running ['C:\\Users\\binstar\\Miniconda\\python.exe', 'C:\\Users\\binstar\\Miniconda\\Scripts\\anaconda-script.py', '--show', '-t', '\\Users\\binstar\\.anaconda.token', 'worker', 'run', '5660bf0cea546b7d099ccc67', '--show-new-procs', '--status-file', '\\Users\\binstar\\worker\\worker.status']

----------------------------------------------------------------------

```

File: binstar_build_client/worker/worker.py [hunk 4]
```diff
@@ -231,7 +231,7 @@ def build(self, job_data):
             build_log.write(msg.encode('utf-8', errors='replace'))
 
             build_log.flush()
-
+            validate_procs()
```
Comment (srossross) [inline binstar_build_client/worker/worker.py]: @PeterDSteinberg maybe you should include this is the build run command not in the worker command. That way the tests will not fail. The script should also have an option to warn and continue.

**Comment by PeterDSteinberg** [comment]
@srossross Sorry I was unclear.  I mean [this build](https://anaconda.org/binstar/binstar-build/builds/419/5) did not have a build worker failure or any problems with the build worker's operation, even though it was on windows worker and [used `conda create`  in the build script](https://anaconda.org/binstar/binstar-build/builds/419/5#L84).  The windows worker was apparently the same build worker that had a test error later (see below) indicating that the `conda create` command shown in previous sentence would have been run on the same conda/python that was doing `anaconda worker run` .  It seems like the current build-binstar-continuum queue is doing `anaconda worker run` and `conda create` on the same conda/python prefix, right?  
```
======================================================================
ERROR: test_build_timeout (binstar_build_client.worker.tests.test_jobs.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\binstar\builds\binstar\binstar-build\env\lib\site-packages\mock\mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "C:\Users\binstar\builds\binstar\binstar-build\source\binstar_build_client\worker\tests\test_jobs.py", line 238, in test_build_timeout
    failed, status = worker.build(job_data)
  File "C:\Users\binstar\builds\binstar\binstar-build\source\binstar_build_client\worker\worker.py", line 234, in build
    validate_procs()
  File "C:\Users\binstar\builds\binstar\binstar-build\source\binstar_build_client\worker\utils\validate_procs.py", line 34, in validate_procs
    "Python prefix: {}".format(" ".join(procs_on_wrong_python)))

BinstarError: There were processes running on the incorrect Python prefix: Pid 1408 is running ['C:\\Users\\binstar\\Miniconda\\python.exe', 'C:\\Users\\binstar\\Miniconda\\Scripts\\anaconda-script.py', '--show', '-t', '\\Users\\binstar\\.anaconda.token', 'worker', 'run', '5660bf0cea546b7d099ccc67', '--show-new-procs', '--status-file', '\\Users\\binstar\\worker\\worker.status']

----------------------------------------------------------------------
```

**Comment by PeterDSteinberg** [comment]
@srossross @pwang @stephenakearns 

To follow up on this issue I did the following experiment:

 * Follow windows build worker instructions up to point of miniconda install
 * Then follow these instructions:

```
set HOME_DIR=C:\Users\anaconda
mkdir %HOME_DIR%\worker
mkdir %HOME_DIR%\.workers
conda config --set always_yes true
conda update conda
conda install pip
conda install -c https://conda.anaconda.org/tpowell unxutils
conda install conda-build
conda install pywin32 psutil mock patch
conda create -n worker --copy python=2.7
activate worker
```

 * The above does `create --copy` then on that environment, install anaconda-build stack
```
REM run chalmers, anaconda-build, clyent, and anaconda-client
REM from develop branch
git clone http://github.com/Anaconda-Server/anaconda-build
cd anaconda-build
git fetch --all 
git checkout fix/test_conda_path
python setup.py develop
cd ..
git clone http://github.com/Anaconda-Server/anaconda-client
cd anaconda-client
python setup.py develop
cd ..
git clone http://github.com/Anaconda-Server/chalmers
cd chalmers
python setup.py develop
cd ..

set WORKER_ID="win-64-psteinberg-aws-3"
set TOKEN_FILE=%HOME_DIR%\.anaconda.token

anaconda auth --create -n "anaconda-build-win-64-psteinberg-99i" --scopes "api:build-worker api:read api:build" --out %TOKEN_FILE%
python -c "import socket;print socket.gethostname()" > %HOME_DIR%\HOSTNAME.txt
set /p HOSTNAME=<%HOME_DIR%\HOSTNAME.txt
anaconda login
anaconda worker register -p win-64 --dist aws-win64 --hostname %HOSTNAME%  psteinberg/abc --name %WORKER_ID%
anaconda logout

REM Configure and run chalmers from the correct python
chalmers stop --all
chalmers remove --all
chalmers add --name worker -c "anaconda --show -t %TOKEN_FILE% worker run %WORKER_ID% --show-new-procs --status-file %HOME_DIR%\worker\worker.status"
chalmers start --all

```

 * Fails at start because the validate_procs() function in this PR detected that chalmers runner was running on the same python prefix.

```
[ERROR]
Traceback (most recent call last):
  File "C:\Users\anaconda\Miniconda\envs\worker\Scripts\anaconda-script.py", lin
e 9, in <module>
    load_entry_point('anaconda-client', 'console_scripts', 'anaconda')()
  File "c:\users\anaconda\anaconda-client\binstar_client\scripts\cli.py", line 8
7, in main
    description=__doc__, version=version)
  File "c:\users\anaconda\anaconda-client\binstar_client\scripts\cli.py", line 6
6, in binstar_main
    return args.main(args)
  File "c:\users\anaconda\anaconda-build\binstar_build_client\worker_commands\ru
n.py", line 30, in main
    validate_procs()
  File "c:\users\anaconda\anaconda-build\binstar_build_client\worker\utils\valid
ate_procs.py", line 34, in validate_procs
    "Python prefix: {}".format(" ".join(procs_on_wrong_python)))
BinstarError: There were processes running on the incorrect Python prefix: Pid 1
28 is running ['C:\\Users\\anaconda\\Miniconda\\envs\\worker\\python.exe', 'c:\\
users\\anaconda\\chalmers\\chalmers\\scripts\\runner.py', 'worker'] Pid 1264 is
running ['C:\\Users\\anaconda\\Miniconda\\envs\\worker\\python.exe', 'C:\\Users\
\anaconda\\Miniconda\\envs\\worker\\Scripts\\chalmers-script.py', 'start', 'work
er']
```
 * If I add the --ignore-process-check flag, the build worker will get started by chalmers okay, then I get build logs [like this](https://anaconda.org/psteinberg/myorg.package1/builds/12/0) that fail in the setup of initial build worker env, but the build-log.txt looks like the build completed. Here is the contents of build-log.txt for the linked build of myorg.package1:

```
[worker] C:\Users\anaconda\builds\psteinberg\myorg.package1>cat build-log.txt
Building on worker WIN-D9IGEGRJCVL (platform win-64)
Starting build psteinberg/myorg.package1#11.1
DIAGNOSTICS
where python
C:\Users\anaconda\Miniconda\envs\worker\python.exe
C:\Users\anaconda\Miniconda\python.exe
where conda
C:\Users\anaconda\Miniconda\Scripts\conda.exe
where pip
C:\Users\anaconda\Miniconda\envs\worker\Scripts\pip.exe
C:\Users\anaconda\Miniconda\Scripts\pip.exe
where anaconda
C:\Users\anaconda\Miniconda\envs\worker\Scripts\anaconda.exe
TASKLIST

Image Name                     PID Session Name        Session#    Mem Usage
========================= ======== ================ =========== ============
System Idle Process              0 Services                   0          4 K
System                           4 Services                   0        276 K
smss.exe                       356 Services                   0      1,060 K
csrss.exe                      504 Services                   0      3,508 K
csrss.exe                      556 Console                    1      3,332 K
wininit.exe                    564 Services                   0      3,708 K
winlogon.exe                   592 Console                    1      5,568 K
services.exe                   652 Services                   0      5,328 K
lsass.exe                      660 Services                   0     10,208 K
svchost.exe                    716 Services                   0     10,064 K
svchost.exe                    744 Services                   0      6,032 K
LogonUI.exe                    848 Console                    1     23,544 K
dwm.exe                        856 Console                    1     24,864 K
svchost.exe                    872 Services                   0     12,192 K
svchost.exe                    904 Services                   0     26,972 K
svchost.exe                    968 Services                   0      8,636 K
svchost.exe                    460 Services                   0     14,728 K
svchost.exe                    496 Services                   0     11,612 K
spoolsv.exe                   1172 Services                   0      9,404 K
LiteAgent.exe                 1200 Services                   0      4,832 K
svchost.exe                   1220 Services                   0      8,120 K
IpOverUsbSvc.exe              1236 Services                   0      9,852 K
sqlservr.exe                  1316 Services                   0    115,272 K
sqlwriter.exe                 1352 Services                   0      5,556 K
svchost.exe                   1368 Services                   0      9,632 K
Ec2Config.exe                 1424 Services                   0     43,484 K
WmiPrvSE.exe                  1556 Services                   0      6,300 K
WmiPrvSE.exe                  1956 Services                   0     15,808 K
fdlauncher.exe                2032 Services                   0      3,480 K
svchost.exe                    440 Services                   0     34,376 K
svchost.exe                   1388 Services                   0      4,396 K
fdhost.exe                    2172 Services                   0      4,440 K
conhost.exe                   2184 Services                   0      2,704 K
msdtc.exe                     3008 Services                   0      6,944 K
csrss.exe                     1592 RDP-Tcp#0                  2     13,856 K
winlogon.exe                  2676 RDP-Tcp#0                  2      5,020 K
dwm.exe                       2796 RDP-Tcp#0                  2     41,668 K
taskhostex.exe                 800 RDP-Tcp#0                  2      8,200 K
rdpclip.exe                    476 RDP-Tcp#0                  2      6,824 K
explorer.exe                  1996 RDP-Tcp#0                  2     96,300 K
cmd.exe                       2452 RDP-Tcp#0                  2      3,660 K
conhost.exe                   2740 RDP-Tcp#0                  2      6,424 K
python.exe                     128 RDP-Tcp#0                  2     24,736 K
conhost.exe                   1264 RDP-Tcp#0                  2      3,872 K
anaconda.exe                  1664 RDP-Tcp#0                  2      1,612 K
python.exe                    1932 RDP-Tcp#0                  2     35,512 K
cmd.exe                       1588 RDP-Tcp#0                  2      2,460 K
tasklist.exe                   828 RDP-Tcp#0                  2      5,648 K
[Setup Build]
Host: WIN-D9IGEGRJCVL
[Setting engine]
conda clean -pt > NUL
conda-clean-build-dir
Conda build root C:\Users\anaconda\conda-bld does not exist
conda clean --lock
type condarc
The system cannot find the file specified.
type condarc
channels:
  - defaults

binstar_upload: no
always_yes: yes
show_channel_urls: yes
Rmdir /s /q "C:\Users\anaconda\builds\psteinberg\myorg.package1\env"
The system cannot find the file specified.
conda create -p "C:\Users\anaconda\builds\psteinberg\myorg.package1\env" --quiet
 --yes python=3
Fetching package metadata: ....
Solving package specifications: ............
Package plan for installation in environment C:\Users\anaconda\builds\psteinberg
\myorg.package1\env:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    python-3.5.0               |                1        37.3 MB  defaults
    setuptools-19.4            |           py35_0         674 KB  defaults
    wheel-0.26.0               |           py35_1          81 KB  defaults
    pip-8.0.1                  |           py35_0         1.6 MB  defaults
    ------------------------------------------------------------
                                           Total:        39.6 MB

The following NEW packages will be INSTALLED:

    pip:        8.0.1-py35_0  defaults
    python:     3.5.0-1       defaults
    setuptools: 19.4-py35_0   defaults
    wheel:      0.26.0-py35_1 defaults

#
# To activate this environment, use:
# > activate C:\Users\anaconda\builds\psteinberg\myorg.package1\env
#
activate C:\Users\anaconda\builds\psteinberg\myorg.package1\env
where conda
C:\Users\anaconda\Miniconda\Scripts\conda.exe
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: No module named 'numpy'
CONDARC C:\Users\anaconda\builds\psteinberg\myorg.package1\condarc
CONDA_PY 35
CONDA_NPY ""

[Fetching Build Source]
The system cannot find the file specified.
dir C:\Users\anaconda\builds\psteinberg\myorg.package1\source.tar.bz2
 Volume in drive C has no label.
 Volume Serial Number is 665F-444D

 Directory of C:\Users\anaconda\builds\psteinberg\myorg.package1

01/28/2016  08:59 PM             5,293 source.tar.bz2
               1 File(s)          5,293 bytes
               0 Dir(s)     680,546,304 bytes free
where python
C:\Users\anaconda\builds\psteinberg\myorg.package1\env\python.exe
C:\Users\anaconda\Miniconda\envs\worker\python.exe
C:\Users\anaconda\Miniconda\python.exe
where tar
C:\Users\anaconda\Miniconda\Scripts\tar.exe
"Extracting Package"
tar jxf C:\Users\anaconda\builds\psteinberg\myorg.package1\source.tar.bz2

[Before Script]
echo "before_script!"
"before_script!"

[Script]
echo "This is my binstar build!"
"This is my binstar build!"

[After Success]
echo "after_success!"
"after_success!"

[After Script]
python myorg/package1/scripts/empty_script.py
Traceback (most recent call last):
  File "myorg/package1/scripts/empty_script.py", line 1, in <module>
    import yaml
ImportError: No module named 'yaml'
Exit BINSTAR_BUILD_RESULT=error
```

**Comment by srossross** [comment]
@PeterDSteinberg this looks like you can break this down into two issue that you can fix:

1. Fix the binstar-build unittests. The tests are not passing, you need to run `python -m unittest discover` and make the unittests pass before you continue.
1. Fix the environment 

     > If I add the --ignore-process-check flag, the build worker will get started by chalmers okay, then I get build logs like this that fail in the setup of initial build worker env, but the build-log.txt looks like the build completed.

    Why are your using the `--ignore-process-check` flag? Doesn't this show that your environment is corrupt. Why do you want to continue before fixing this.

1. Fix the build. Does your build run on a posix machine?  What does this statement mean?
    ```dir C:\Users\anaconda\builds\psteinberg\myorg.package1\source.tar.bz2
         Volume in drive C has no label.
         Volume Serial Number is 665F-444D
    ```

**Comment by PeterDSteinberg** [comment]
@srossross 

Regarding your enumerated comments (in order above):

 *  The `python -m unittest discover` is passing all tests on there now.
 *  It seems like the `validate_procs` is overactive because it is detecting `chalmers start --all` or `chalmers start worker` as invalid, running on the wrong python (in the quoted error above).  I mean that it seems like there is always going to be a chalmers process running on the same  environment while the process doing `worker run` is running `validate_procs` since chalmers is doing it in a subprocess.  I have tried various permuations in the `chalmers add` program command quoted above to refer to the `chalmers-script.py` in the worker env and `anaconda-script.py` in the worker env and none of those permutations allow me to get past the worker's `validate_procs`.  The processes I have seen in conflict include:
  * chalmers `runner.py` is also running on the worker env's python.exe
  * chalmers `start --all` is also running on the worker env's python.exe

*  `dir` was an extra diagnostic statement I made in build_script.bat template and I have removed it.

**Comment by srossross** [comment]
> chalmers start --all or chalmers start worker as invalid, running on the wrong python (in the quoted error above). I mean that it seems like there is always going to be a chalmers process running on the same environment while the process doing worker run is running validate_procs since chalmers is doing it in a subprocess.

I don't think this should be the case `chalmers start worker` should always run in with  same executable as `chalmers start --all` [launch code](https://github.com/Anaconda-Server/chalmers/blob/master/chalmers/program/nt.py#L52-L57) And you have control of where you run `chalmers start --all` from

File: binstar_build_client/worker_commands/run.py [hunk 5]
```diff
@@ -25,7 +26,8 @@ def main(args):
     args.conda_build_dir = args.conda_build_dir.format(platform=worker_config.platform)
     log.info("Using conda build directory: {}".format(args.conda_build_dir))
     log.info(str(worker_config))
-
+    if not args.ignore_process_check:
+        validate_procs()
```
Comment (srossross) [inline binstar_build_client/worker_commands/run.py]: This should still run `validate_procs` but only log the information

**Comment by PeterDSteinberg** [comment]
@srossross I tried again, but instead of `conda create --copy` or `xcopy` to create a separate build env, I did a second Miniconda installation, via the one-click exe, at C:\Users\anaconda\ChalmersMiniconda.  Now it is no longer failing the runtime check that is this PR (check for conflicting python processes is okay now).  Python 3 build workers are now sometimes able to initialize state, but still having trouble with Python 2 and sometimes Python 3 with `conda create` in the build script (I also experimented with changing the `conda create` in the build_script.bat to `conda create --copy` but that didn't improve Python 2 build setup issues).  Here is what I am doing:
```
REM testing on ec2-52-90-136-86.compute-1.amazonaws.com
REM followed these deployment instructions:
REM        https://sites.google.com/a/continuum.io/wiki/design-and-development-documents/binstar/binstar-build-win-64-config
REM up through the point of installing Miniconda
REM in C:\Users\anaconda\Miniconda
REM Then ran the commands below:

set HOME_DIR=C:\Users\anaconda
mkdir %HOME_DIR%\worker
mkdir %HOME_DIR%\.workers
conda config --set always_yes true
conda update conda
conda install pip
conda install -c https://conda.anaconda.org/tpowell unxutils
conda install conda-build
conda install pywin32 psutil mock patch


REM Now install Miniconda again at C:\Users\anaconda\ChalmersMiniconda


set WORKER_ID="win-64-psteinberg-aws-3"
set TOKEN_FILE=%HOME_DIR%\.anaconda.token
set BPY=%HOME_DIR%\ChalmersMiniconda\python.exe
set BC=%HOME_DIR%\ChalmersMiniconda\Scripts\conda-script.py
set AC=%HOME_DIR%\ChalmersMiniconda\Scripts\anaconda-script.py
set CH=%HOME_DIR%\ChalmersMiniconda\Scripts\chalmers-script.py

%BPY% %BC% config --set always_yes true
%BPY% %BC% update conda
%BPY% %BC% install pip
%BPY% %BC% install -c https://conda.anaconda.org/tpowell unxutils
%BPY% %BC% install conda-build
%BPY% %BC% install pywin32 psutil mock patch

git clone http://github.com/Anaconda-Server/anaconda-build
cd anaconda-build
git fetch --all
git checkout fix/test_conda_path
%BPY% setup.py develop
cd ..
git clone http://github.com/Anaconda-Server/anaconda-client
cd anaconda-client
%BPY% setup.py develop
cd ..
git clone http://github.com/Anaconda-Server/chalmers
cd chalmers
%BPY% setup.py develop
cd ..

%BPY% %AC% auth --create -n "anaconda-build-win-64-psteinberg-99l" --scopes "api:build-worker api:read api:build" --out %TOKEN_FILE%
%BPY% -c "import socket;print socket.gethostname()" > %HOME_DIR%\HOSTNAME.txt
set /p HOSTNAME=<%HOME_DIR%\HOSTNAME.txt
%BPY% %AC% login
%BPY% %AC% worker register -p win-64 --dist aws-win64 --hostname %HOSTNAME%  psteinberg/abc --name %WORKER_ID%
%BPY% %AC% logout

set HOME_PY=%HOME_DIR%\Miniconda\python.exe
set HOME_AC=%HOME_DIR%\Miniconda\Scripts\anaconda-script.py

REM Configure and run chalmers from the correct python
%BPY% %CH% stop --all
%BPY% %CH% remove --all
%BPY% %CH% add --name worker -c "%HOME_PY% %HOME_AC% --show -t %TOKEN_FILE% worker run %WORKER_ID% --show-new-procs --status-file %HOME_DIR%\worker\worker.status --ignore-process-check"
%BPY% %CH% start --all
```

**Comment by srossross** [comment]
@PeterDSteinberg how close is this to being ready to merge?

**Comment by PeterDSteinberg** [comment]
@srossross This is ready to merge.

File: binstar_build_client/worker/utils/data/build_script.bat [hunk 6]
```diff
@@ -209,6 +209,8 @@ goto:eof
 
     echo Rmdir /s /q "%BUILD_ENV_PATH%"
     Rmdir /s /q "%BUILD_ENV_PATH%"
+    echo "Check binstar-build-validate-procs before build script"
```
Comment (srossross) [inline binstar_build_client/worker/utils/data/build_script.bat]: This does not belong here. The check only needs to happen on startup.

File: binstar_build_client/worker/utils/validate_procs.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    for proc in psutil.process_iter():
+        if proc.pid == my_pid:
+            continue
+        cmd = None
+        try:
+            cmd = proc.cmdline()
+        except psutil.AccessDenied:
+            continue
+        except psutil.ZombieProcess:
+            log.info('ZombieProcess: {} {}'.format(proc.pid, cmd))
```
Comment (srossross) [inline binstar_build_client/worker/utils/validate_procs.py]: Can you add comments here?

Comment (srossross) [inline binstar_build_client/worker/utils/validate_procs.py]: Why are you not logging the psutil.AccessDenied error too? 

Also, there is no continue statement here? Is that a bug? can you add a test to show that this the exceptions are handled correctly?

File: binstar_build_client/worker/utils/validate_procs.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        open_files = ['Function_Not_Available']
+        try:
+            if hasattr(proc, 'open_files'):
+                open_files = proc.open_files()
+        except psutil.AccessDenied:
+            open_files = ['AccessDenied']
+        msg = "Pid {} is running {} with open files {}"
+        if cmd and cmd[0].strip().startswith(executable_dir):
```
Comment (srossross) [inline binstar_build_client/worker/utils/validate_procs.py]: why not use `proc.exe()` instead of `proc.cmdline()` above?

File: binstar_build_client/worker_commands/run.py [hunk 9]
```diff
@@ -25,6 +27,15 @@ def main(args):
     args.conda_build_dir = args.conda_build_dir.format(platform=worker_config.platform)
     log.info("Using conda build directory: {}".format(args.conda_build_dir))
     log.info(str(worker_config))
+    procs_on_wrong_python = validate_procs()
+    if not args.ignore_process_check and procs_on_wrong_python:
```
Comment (srossross) [inline binstar_build_client/worker_commands/run.py]: should you just pass `args.ignore_process_check` as an argument to `validate_procs`?

File: binstar_build_client/worker/utils/validate_procs.py (hunk 7 above)
Comment (PeterDSteinberg) [inline binstar_build_client/worker/utils/validate_procs.py]: @srossross  I will add comments and fix continue statement.  I will log AccessDenied

File: binstar_build_client/worker/utils/validate_procs.py (hunk 8 above)
Comment (PeterDSteinberg) [inline binstar_build_client/worker/utils/validate_procs.py]: On Windows `proc.exe()` is simply `"python.exe"` not the full path.

File: binstar_build_client/worker_commands/run.py (hunk 9 above)
Comment (PeterDSteinberg) [inline binstar_build_client/worker_commands/run.py]: Yes I'll do that.

File: binstar_build_client/worker/utils/data/build_script.bat (hunk 6 above)
Comment (PeterDSteinberg) [inline binstar_build_client/worker/utils/data/build_script.bat]: @srossross I thought we also wanted something in the build script right before `conda create` that is causing problems (in addition to at startup)?

File: binstar_build_client/worker/utils/validate_procs.py (hunk 8 above)
Comment (srossross) [inline binstar_build_client/worker/utils/validate_procs.py]: [image: Inline image 1]

Comment (PeterDSteinberg) [inline binstar_build_client/worker/utils/validate_procs.py]: I stand corrected.  Will fix.

File: binstar_build_client/worker/utils/data/build_script.bat (hunk 6 above)
Comment (srossross) [inline binstar_build_client/worker/utils/data/build_script.bat]: @PeterDSteinberg I am not sure that this is required.  This is only to sanitize the initial build configuration. There is nothing more a user can do. The next changes need to go into conda directly to report exactly why it is failing.

Comment (srossross) [inline binstar_build_client/worker/utils/data/build_script.bat]: Also, won't this make builds fail even if a user runs the worker with `ignore_process_check`?

Comment (PeterDSteinberg) [inline binstar_build_client/worker/utils/data/build_script.bat]: @srossross  Ok.  Removing it from here and will look at putting that check into `conda create` source.

**Comment by PeterDSteinberg** [comment]
I modified the snippet of deployment code above as follows

```
set HOME_PY=%HOME_DIR%\Miniconda\python.exe
set HOME_AC=%HOME_DIR%\Miniconda\Scripts\anaconda-script.py

REM Configure and run chalmers from the correct python
%BPY% %CH% stop --all
%BPY% %CH% remove --all
%BPY% %CH% add --name worker -c "%HOME_PY% %HOME_AC% --show -t %TOKEN_FILE% worker run %WORKER_ID% --show-new-procs --status-file %HOME_DIR%\worker\worker.status "
%BPY% %CH% start --all
```

And I made a [PR in conda](https://github.com/conda/conda/pull/2023/) for more diagnostic printing on permissions errors on windows, adding `tasklist` call and psutil process iteration to print out any process that may be holding onto the file with permissions errors.

In [my build of myorg.package1](https://anaconda.org/psteinberg/myorg.package1/builds/54/0) which has the windows permission error, I see in the diagnostic printing of tasklist [here](https://anaconda.org/psteinberg/myorg.package1/builds/54/0#L107) and [here](https://anaconda.org/psteinberg/myorg.package1/builds/54/0#L110) that there are two not-responding python process, including the `anaconda worker run` as one of the unresponsive processes.

**Comment by srossross** [comment]
:+1:

**Comment by stephenakearns** [comment]
This PR is in support of issue #150 - Fix windows builds

**Comment by PeterDSteinberg** [comment]
@srossross As described in the final comment on #150, I think this runtime check in this PR is over-active and should not be merged.

**Comment by dsludwig** [comment]
@PeterDSteinberg if this should not be merged, do you want to close the PR?

**Comment by srossross** [comment]
> @srossross As described in the final comment on #150, I think this runtime check in this PR is over-active and should not be merged.

@PeterDSteinberg I am not sure I understand why this was closed. can you summarize?

**Comment by PeterDSteinberg** [comment]
@srossross Even with an ideal two-Miniconda set up (one at `C:\Miniconda` and one at `C:\Users\binstar\Miniconda`) and running chalmers on the Admin python, this runtime check still fails because chalmers runner.py is on the same python as `anaconda worker`.  I think it is overactive because in ca. one week of testing with it, I have not found a way to install two Miniconda's and configure chalmers in a way that does not trip this runtime check.

**Comment by srossross** [comment]
`runner.py is on the same python as anaconda worker` the anaconda worker should be on the `C:\Miniconda ` right? Not on `C:\Users\binstar\Miniconda`

**Comment by PeterDSteinberg** [comment]
@srossross I'll merge this PR into develop and copy the full error message here later this afternoon.

**Comment by srossross** [comment]
@PeterDSteinberg I'm not sure what is going on here.  I don't know if this should be merged or not. but I don't understand what the issue is.

**Comment by PeterDSteinberg** [comment]
@srossross Sorry I was just about to edit that message.  What I meant was that in a local branch I will merge this PR into develop so that I can get the exact quote of the error message.

**Comment by srossross** [comment]
OK, can you confirm that the worker is not running  in `C:\Users\binstar\Miniconda`? This should be running under C:`\Miniconda`.
